src/uu/head/src/parse.rs

```diff
@@ -0,0 +1,237 @@
+}
+/// Parses obsolete syntax
+/// head -NUM
+/// (note:)
+pub fn parse_obsolete(src: &str) -> Option<Result<(usize, bool), ParseError>> {
+    let re = regex::Regex::new(r"^-(\d+)(?:([cbkm])+)?$").unwrap();
+    match re.captures(src) {
+        Some(cap) => {
+            let num = cap.get(1).unwrap();
+            let mut num = match num.as_str().parse::<usize>() {
+                Ok(n) => n,
+                Err(_) => {
+                    return Some(Err(ParseError::Overflow));
+                }
+            };
+            let mut bytes = false;
+            if let Some(m) = cap.get(2) {
+                let newn = match m.as_str().chars().next().unwrap_or(0 as char) {
+                    'c' => num.checked_mul(1),
+                    'b' => num.checked_mul(512),
+                    'k' => num.checked_mul(1024),
+                    'm' => num.checked_mul(1024 * 1024),
+                    _ => unreachable!(),
+                };
+                match newn {
+                    Some(n) => {
+                        num = n;
+                        bytes = true;
+                    }
+                    None => return Some(Err(ParseError::Overflow)),
+                }
+            }
+            Some(Ok((num, bytes)))
+        }
+        None => None,
+    }
+}
+/// Parses an -c or -n argument,
+/// the bool specifies whether to read from the end
+pub fn parse_num(src: &str) -> Result<(usize, bool), ParseError> {
+    // we're not using the bytefmt or byte_unit crates
+    // because they aren't fully `head` format compliant
+
+    if src.is_empty() {
+        return Err(ParseError::Syntax);
+    }
+
+    // (1) is the number
+    // (2) is b
+    // (3) is any of the other units
+    // (4) is `i`
+    // (5) is `B`
+    let re = regex::Regex::new(r"^-?(\d+)?(?:(b)|([kKmMgGtTpPeEzZyY])(?:(i)B|(B))?)?$").unwrap();
+
+    match re.captures(src) {
+        Some(cap) => {
+            let number = match cap.get(1) {
+                Some(n) => {
+                    // we're matching digits so it's ok to unwrap
+                    n.as_str().parse().unwrap()
+                }
+                None => 1usize,
+            };
+            //`{unit}iB` means base2 as well as just `{unit}`
+            let is_base_2 = cap.get(4).is_some() || cap.get(5).is_none();
+
+            let multiplier = if cap.get(2).is_some() {
+                512
+            } else if let Some(unit) = cap.get(3) {
+                //this is safe as the first byte cannot be
+                //part of a multi-byte character if it is
+                //an ascii character
+                match &unit.as_str().as_bytes()[0] {
+                    b'k' | b'K' => {
+                        if is_base_2 {
+                            1024u128.pow(1)
+                        } else {
+                            1000u128.pow(1)
+                        }
+                    }
+                    b'm' | b'M' => {
+                        if is_base_2 {
```

Comment:
Since all of these branches are the same, restructure it like so:
```rust
let exp = match blah {
    b'k' | b'K' => 1,
    b'm' | b'M' => 2,
    // and so on...
};

if is_base_2 {
    1024u128.pow(exp)
} else {
    1000u128.pow(exp)
}
```

I think there's also `.to_ascii_lowercase()` or something function you can use to avoid checking both the lowercase and uppercase versions of the letters.